src/expr/src/scalar/mod.rs

```diff
@@ -890,6 +897,114 @@ impl MirScalarExpr {
+                    for list_create in &mut list_creates {
+                        let list_create_args = match list_create {
+                            MirScalarExpr::CallVariadic {
+                                func: VariadicFunc::ListCreate { .. },
+                                exprs,
+                            } => exprs,
+                            _ => unreachable!(), // func cannot be anything else than a ListCreate
+                        };
+                        // ListIndex gives null on an out-of-bounds index
+                        if index_i64 >= 0 && index_i64 < list_create_args.len().try_into().unwrap()
+                        {
+                            let index: usize = index_i64.try_into().unwrap();
+                            **list_create = list_create_args.swap_remove(index);
+                        } else {
+                            let typ = list_create_type(list_create);
+                            **list_create = MirScalarExpr::literal_null(typ);
+                        }
+                    }
+                } else {
+                    // We can't remove this index.
+                    // Increment i only in this branch, because the other removes
+                    // the current element of index_exprs.
+                    i += 1;
+                    // Change list_creates to all the arguments of all the list_creates.
+                    list_creates = list_creates
+                        .drain(..)
+                        .flat_map(|list_create| match list_create {
+                            MirScalarExpr::CallVariadic {
+                                func: VariadicFunc::ListCreate { .. },
+                                exprs: list_create_args,
+                            } => list_create_args,
+                            // func cannot be anything else than a ListCreate
+                            _ => unreachable!(),
+                        })
+                        .collect();
+                }
+            }
+            if index_exprs.len() == 0 {
+                assert_eq!(list_creates.len(), 1);
+                // All indexes have been removed.
+                // Remove the ListIndex, because it would have 0 arguments.
+                // (list_creates[0] is not necessarily a ListCreate at this point.)
+                list_creates[0].take()
+            } else {
+                // Either some of the indexes were not literals, or the nested ListCreates stopped
+                // actually being ListCreates when we reached a certain depth.
+                // Put together the remaining ListCreates and the remaining indexes.
+                let mut exprs: Vec<MirScalarExpr> = vec![top_list_create.clone()];
+                exprs.append(index_exprs);
+                MirScalarExpr::CallVariadic {
+                    func: VariadicFunc::ListIndex,
+                    exprs,
+                }
```

Comment:
What is happening in this function is `top_list_create` is being modified in place, and `list_create` is a list of mutable references to different subexpressions in `top_list_create`. To make this fact clearer, I propose:
* replacing both `list_creates[0].take()` and `top_list_create.clone()` in this section with `top_list_create.take()`.
* Simplify the comments here to something like: "If all list indexes have been evaluated, return the reduced expression. Otherwise, rebuild the `ListIndex` call."
* renaming `list_creates` to `cursors` or `mut_refs` to convey that it is a list of references and not a list of expression objects.
* possibly rename `top_list_create` to something like `list_create_to_reduce` to save the effort of clarifying that it might not be a ListCreate anymore.

Reply:
Mostly done, except:
- Actually, the `take` in `top_list_create.take()` is not even needed.
- Renamed `list_creates` to `list_create_mut_refs` to also not lose the info that is in the original name.